### ISSUE TITLE
Discrepancy in Dropout between DTensor and torch.Tensor

### ISSUE THREAD

**Issue by bonpyt** [pytorch/pytorch#146190]
### 🐛 Describe the bug

We are seeing differences in Dropout between `DTensor` and `torch.Tensor` and we think this is due to the CUDA PRNG state changing when using `DTensor`, but not when using `torch.Tensor`.
```
#!/usr/bin/env python3

import torch
import os

from contextlib import nullcontext
from torch.distributed.tensor import DTensor
from torch.distributed._tensor.device_mesh import init_device_mesh


def print_prng():
    print(f"torch PRNG:  {torch.get_rng_state().sum()}")
    print(f"CUDA  PRNG:  {torch.cuda.get_rng_state().sum()}")


def main():
    distributed = os.environ.get("RANK") is not None
    print(f"distributed: {distributed}")

    with init_device_mesh(
        device_type="cuda", mesh_shape=(1,)
    ) if distributed else nullcontext() as device_mesh:

        seed = 42
        torch.manual_seed(seed)
        torch.cuda.manual_seed(seed)

        dropout = torch.nn.Dropout()

        print_prng()
        tensor = torch.rand(10, 10)
        print_prng()

        if distributed:
            tensor = DTensor.from_local(tensor)
            tensor = tensor.full_tensor()

        print(f"tensor:  {tensor.sum().item()}")
        print_prng()
        print(f"dropout: {dropout(tensor).sum().item()}")
        print_prng()
        print(f"dropout: {dropout(tensor).sum().item()}")
        print_prng()


if __name__ == "__main__":
    main()
```

```
$ ./test_dropout.py 
distributed: False
torch PRNG:  316607
CUDA  PRNG:  42
torch PRNG:  319252
CUDA  PRNG:  42
tensor:  51.582977294921875
torch PRNG:  319252
CUDA  PRNG:  42
dropout: 63.54338455200195
torch PRNG:  319252
CUDA  PRNG:  42
dropout: 53.90919494628906
torch PRNG:  319252
CUDA  PRNG:  42

$ torchrun --nnodes=1 --nproc_per_node=1 ./test_dropout.py 
distributed: True
torch PRNG:  316607
CUDA  PRNG:  42
torch PRNG:  319252
CUDA  PRNG:  42
tensor:  51.582977294921875
torch PRNG:  319252
CUDA  PRNG:  42
dropout: 52.835205078125
torch PRNG:  319252
CUDA  PRNG:  46
dropout: 40.78480529785156
torch PRNG:  319252
CUDA  PRNG:  50
```

### Versions

```
Collecting environment information...
PyTorch version: 2.6.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A
OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: 10.0.0-4ubuntu1 
CMake version: version 3.16.3
Libc version: glibc-2.31
Python version: 3.12.8 (main, Dec  4 2024, 08:54:13) [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-116-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA H200
GPU 1: NVIDIA H200
GPU 2: NVIDIA H200
GPU 3: NVIDIA H200
GPU 4: NVIDIA H200
GPU 5: NVIDIA H200
GPU 6: NVIDIA H200
GPU 7: NVIDIA H200
Nvidia driver version: 535.216.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Byte Order:                           Little Endian
Address sizes:                        46 bits physical, 57 bits virtual
CPU(s):                               96
On-line CPU(s) list:                  0-95
Thread(s) per core:                   1
Core(s) per socket:                   48
Socket(s):                            2
NUMA node(s):                         2
Vendor ID:                            GenuineIntel
CPU family:                           6
Model:                                207
Model name:                           INTEL(R) XEON(R) PLATINUM 8568Y+
Stepping:                             2
CPU MHz:                              2300.000
BogoMIPS:                             4600.00
L1d cache:                            4.5 MiB
L1i cache:                            3 MiB
L2 cache:                             192 MiB
L3 cache:                             600 MiB
NUMA node0 CPU(s):                    0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38,40,42,44,46,48,50,52,54,56,58,60,62,64,66,68,70,72,74,76,78,80,82,84,86,88,90,92,94
NUMA node1 CPU(s):                    1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39,41,43,45,47,49,51,53,55,57,59,61,63,65,67,69,71,73,75,77,79,81,83,85,87,89,91,93,95
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Vulnerable
Vulnerability Spectre v1:             Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:             Vulnerable; IBPB: disabled; STIBP: disabled; PBRSB-eIBRS: Vulnerable; BHI: Vulnerable
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 invpcid_single cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
Versions of relevant libraries:
[pip3] mypy==1.13.0
[pip3] mypy-extensions==1.0.0
[pip3] mypy-protobuf==3.6.0
[pip3] numpy==2.0.1
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-cusparselt-cu12==0.6.2
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] pytorch-lightning==2.4.0
[pip3] torch==2.6.0
[pip3] torchmetrics==1.6.0
[pip3] torchvision==0.19.1
[pip3] triton==3.2.0
[pip3] tritonclient==2.46.0
[conda] Could not collect
```

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

**Comment by bonpyt** [comment]
Closing as the difference is caused by a difference in devices.
Using the device mesh with `device_type='cuda'` puts the DTensor on the GPU and therefore obviously the CUDA PRNG state changes with a dropout call.
In the non-DTensor case we never actually moved the tensors to the GPU so the CUDA PRNG state never changes and we end up with a difference.
